### PR TITLE
Revamp LINES

### PR DESCRIPTION
**Pull request by paulapatience**
- Keep empty lines.
- Handle more EOL characters (CRLF in particular is tricky).
- Optionally keep EOL characters in the lines.

Although keeping empty lines is a backwards-incompatible change, I would argue that omitting empty lines from the result is a bug. For instance, it prevents you from being able to count the lines in a file.

I failed in getting docs.lisp to update REFERENCE.md.

### CODE REVIEW

File: strings.lisp [hunk 1]
```diff
@@ -308,17 +308,187 @@ removed."
...
-(defun lines (string)
-  "A list of lines in STRING."
-  (declare (string string))
-  (values (split-sequence-if #'newline? string :remove-empty-subseqs t)))
+(-> lines ((or null string)
+           &key
+           (:eol-style (or (member nil :cr :lf :crlf :ascii :unicode)
+                           function))
+           (:honor-crlf boolean)
+           (:keep-eols boolean))
```
[the remark this answers is not in the document]
Comment (paulapatience) [reply strings.lisp:316]: You are right. I will fix this.

File: strings.lisp [hunk 2]
```diff
@@ -308,17 +308,187 @@ removed."
...
+
+  (delete-if #'uiop:emptyp (serapeum:lines string :eol-style :unicode))
+
+To additionally omit lines consisting only of whitespace:
+
+  (delete-if #'uiop:emptyp
+             (mapcar #'serapeum:trim-whitespace
+                     (serapeum:lines string :eol-style :unicode)))"
+  ;; Ensure that we split ASCII and Unicode strings predictably.
+  #.(ecase #\Newline ((#\Return #\Linefeed)))
```
[the remark this answers is not in the document]
Comment (paulapatience) [reply strings.lisp:429]: Actually, we could remove it entirely. The docstring already mentions that :CR, :LF, :CRLF, :ASCII, and :UNICODE require the Common Lisp implementation to use ASCII or Unicode, so we could amend it to say that only for those EOL-STYLEs must #\Newline be either #\Return or #\Linefeed (which is likely the case on any such implementation). The NIL and predicate EOL-STYLEs do not require the implementation to use ASCII or Unicode at all. What do you think?

[the remark this answers is not in the document]
Comment (ruricolist) [reply strings.lisp:429]: I'm in favor of removing the check altogether and amending the documentation.

**Comment by paulapatience** [comment]
I made the changes you requested and additionally embellished the examples.

**Pull request merged**